prober/tcp.go

```diff
@@ -146,19 +145,14 @@ func ProbeTCP(ctx context.Context, target string, module config.Module, registry
 	for i, qr := range module.TCP.QueryResponse {
 		level.Info(logger).Log("msg", "Processing query response entry", "entry_number", i)
 		send := qr.Send
-		if qr.Expect != "" {
-			re, err := regexp.Compile(qr.Expect)
-			if err != nil {
-				level.Error(logger).Log("msg", "Could not compile into regular expression", "regexp", qr.Expect, "err", err)
-				return false
-			}
+		if qr.Expect.Regexp != nil && qr.Expect.Regexp.String() != "" {
```

Comment:
Isn't Regexp always non-nil?

Reply:
> Isn't Regexp always non-nil?

There is the case that config.Module instance is created in `prober/tcp_test.go`.
https://github.com/prometheus/blackbox_exporter/blob/49ae740bd6f5626585cef994c294bba00a5ff02a/prober/tcp_test.go#L359-L378

And when `Expect` isn't explicitly specified as following line, it's possible that `qr.Expect.Rege` is nil.
https://github.com/prometheus/blackbox_exporter/blob/49ae740bd6f5626585cef994c294bba00a5ff02a/prober/tcp_test.go#L364

But that occors only in the test senario. In this case, should I modify `prober/tcp_test.go` as following.

```
   {Expect: config.MustNewRegexp(""), Send: "EHLO tls.prober"}, 
```

Could you give me an advice.

Reply:
You just need the first predicate here, only check for nil. Expecting an empty response is valid after all.

Reply:
> You just need the first predicate here, only check for nil.
Thank you. I deleted the latter predicate.